groups/reconcile.go

```diff
@@ -0,0 +1,405 @@
+	if err != nil {
+		return fmt.Errorf("unable to retrieve users in domain: %v", err)
+	}
+	if len(g.Groups) == 0 {
+		log.Println("No groups found.")
+		return nil
+	}
+	for _, g := range g.Groups {
+		// Don't touch existing mailing lists, we should
+		// always prefix with "k8s-infra-"
+		if !strings.HasPrefix(g.Email, "k8s-infra-") {
+			continue
+		}
+		found := false
+		for _, g2 := range groupsConfig.Groups {
+			if g2.EmailId == g.Email {
+				found = true
+				break
+			}
+		}
+		if found {
+			continue
+		}
+		// We did not find the group in our groups.xml, so delete the group
+		if config.DryRun {
+			log.Printf("dry-run : Skipping removing group %s\n", g.Email)
+		} else {
+			err := service.Groups.Delete(g.Email).Do()
+			if err != nil {
+				return fmt.Errorf("unable to remove group %s : %v", g.Email, err)
+			} else {
+				log.Printf("Removing group %s\n", g.Email)
+			}
+		}
+
+	}
+	return nil
+}
+
+func updateGroupSettingsToAllowExternalMembers(srv *groupssettings.Service, groupEmailId string) error {
+	g2, err := srv.Groups.Get(groupEmailId).Do()
+	if err != nil {
+		if apierr, ok := err.(*googleapi.Error); ok && apierr.Code == http.StatusNotFound {
+			log.Printf("skipping updating group settings as group %s has not yet been created\n", groupEmailId)
+			return nil
+		}
+		return fmt.Errorf("unable to retrieve group info for group %s: %v", groupEmailId, err)
+	}
+	if g2.AllowExternalMembers != "true" {
+		if config.DryRun {
+			log.Printf("dry-run : skipping updating group settings\n")
+		} else {
+			_, err := srv.Groups.Patch(groupEmailId, &groupssettings.Groups{
+				AllowExternalMembers: "true",
+			}).Do()
+			if err != nil {
+				return fmt.Errorf("unable to update group info for group %s: %v", groupEmailId, err)
+			} else {
+				log.Printf("> Successfully updated group settings for %s to allow external members\n", groupEmailId)
+			}
+		}
+	}
+	return nil
+}
+
+func addMembersToGroup(service *admin.Service, groupEmailId string, members []string) error {
+	l, err := service.Members.List(groupEmailId).Do()
+	if err != nil {
+		if apierr, ok := err.(*googleapi.Error); ok && apierr.Code == http.StatusNotFound {
+			log.Printf("skipping adding members to group %s as it has not yet been created\n", groupEmailId)
+			return nil
+		}
+		return fmt.Errorf("unable to retrieve members in group %s: %v", groupEmailId, err)
+	}
+
+	for _, m := range members {
+		found := false
+		if len(l.Members) > 0 {
+			for _, m2 := range l.Members {
+				if m2.Email == m {
+					found = true
+					break
+				}
+			}
+		}
+		if found {
+			continue
+		}
+		// We did not find the person in the google group, so we add them
+		if config.DryRun {
+			log.Printf("dry-run : Skipping adding %s to %s\n", m, groupEmailId)
+		} else {
+			_, err := service.Members.Insert(groupEmailId, &admin.Member{
+				Email: m,
+				Role:  "MEMBER",
+			}).Do()
+			if err != nil {
+				return fmt.Errorf("unable to add %s to %s : %v", m, groupEmailId, err)
+			} else {
```

Comment:
else is not needed.

https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow

Reply:
Done